File: server/src/db/devices.py
from typing import Optional

from sqlite3 import Cursor

from src.db.util import read_sql_init_file
from src.models import (
    Device,
)
from src.logging import get_logger

logger = get_logger(__name__)


def init_devices(cursor: Cursor):
    logger.info("Initializing devices table")
    init_query = read_sql_init_file("devices")
    try: 
        cursor.execute(init_query)
    except Exception as e:
        logger.error('Failed to initialize devices table', {'error': str(e)})
        raise e


def create_device(cursor: Cursor, device: Device):
    logger.info('Creating device', {'device': device.model_dump()})
    try:
        cursor.execute(
            """
            INSERT INTO devices (mac, ip, name)
            VALUES (?, ?, ?)
            """,
            (device.mac, device.ip, device.name),
        )
    except Exception as e:
        logger.error('Failed to create device', {'device': device.model_dump(), 'error': str(e)})
        raise e
    device.id = cursor.lastrowid
    logger.debug('Device created', {'device': device.model_dump()})


def update_device(cursor: Cursor, device: Device):
    logger.info('Updating device', {'device': device.model_dump()})
    try:
        cursor.execute(
            """
            UPDATE devices
            SET mac = ?, ip = ?, name = ?
            WHERE id = ?
            """,
            (device.mac, device.ip, device.name, device.id),
        )
    except Exception as e:
        logger.error('Failed to update device', {'device': device.model_dump(), 'error': str(e)})
        raise e
    logger.debug('Device updated', {'device': device.model_dump()})


def record_ping(cursor: Cursor, device_id: int):
    logger.debug('Recording ping', {'device_id': device_id})
    try:
        cursor.execute(
            """
            UPDATE devices
            SET last_ping = CURRENT_TIMESTAMP
            WHERE id = ?
            """,
            (device_id,),
        )
    except Exception as e:
        logger.error('Failed to record ping', {'device_id': device_id, 'error': str(e)})
        raise e
    logger.debug('Ping recorded', {'device_id': device_id})


def delete_device(cursor: Cursor, device_id: int):
    logger.info('Deleting device', {'device_id': device_id})
    try: 
        cursor.execute(
            """
            DELETE FROM devices WHERE id = ?
            """,
            (device_id,),
        )
    except Exception as e:
        logger.error('Failed to delete device', {'device_id': device_id, 'error': str(e)})
        raise e
    logger.debug('Device deleted', {'device_id': device_id})


def find_device_by_id(cursor: Cursor, device_id: int) -> Device | None:
    logger.debug('Finding device by id', {'device_id': device_id})
    try:
        cursor.execute(
            """
            SELECT * FROM devices WHERE id = ?
            """,
            (device_id,),
        )
        device = cursor.fetchone()
    except Exception as e:
        logger.error('Failed to find device by id', {'device_id': device_id, 'error': str(e)})
        raise e
    logger.debug('Device found', {'device_data': device})
    try:
        if device:
            return Device(**device)
        return None
    except Exception as e:
        logger.error('Failed to parse device object', {'device_data': device, 'error': str(e)})
        raise e


def find_by_mac(cursor: Cursor, mac: str) -> Device | None:
    logger.debug('Finding device by mac', {'mac': mac})
    try:
        cursor.execute(
            """
            SELECT * FROM devices WHERE mac = ?
            """,
            (mac,),
        )
        device = cursor.fetchone()
    except Exception as e:
        logger.error('Failed to find device by mac', {'mac': mac, 'error': str(e)})
        raise e
    logger.debug('Device found', {'device_data': device})
    try:
        if device:
            return Device(**device)
        return None
    except Exception as e:
        logger.error('Failed to parse device object', {'device_data': device, 'error': str(e)})
        raise e


def find_by_name(cursor: Cursor, name: str) -> Device | None:
    logger.debug('Finding device by name', {'name': name})
    try:
        cursor.execute(
            """
            SELECT * FROM devices WHERE name = ?
            """,
            (name,),
        )
        device = cursor.fetchone()
    except Exception as e:
        logger.error('Failed to find device by name', {'name': name, 'error': str(e)})
        raise e
    logger.debug('Device found', {'device_data': device})
    try:
        if device:
            return Device(**device)
        return None
    except Exception as e:
        logger.error('Failed to parse device object', {'device_data': device, 'error': str(e)})


def list_devices(cursor: Cursor, room: Optional[int] = None, connected: Optional[bool] = None):
    logger.debug('Listing devices', {'room': room, 'connected': connected})
    query = "SELECT * FROM devices"
    where_clauses = []
    args = []
    if room == 0: # room_id=0 means no room
        where_clauses.append("room_id IS NULL")
    elif room is not None:
        where_clauses.append("room_id = ?")
        args.append(room)
    if connected is not None:
        where_clauses.append("connected = ?")
        args.append(connected)
    if where_clauses:
        query = query + " Where " + ", ".join(where_clauses)
    try: 
        cursor.execute(query, args)
        devices = cursor.fetchall()
    except Exception as e:
        logger.error('Failed to list devices', {'room': room, 'connected': connected, 'query': query, 'error': str(e)})
        raise e
    logger.debug('Devices listed', {'devices': devices})
    try: 
        if devices:
            return [Device(**device) for device in devices]
    except Exception as e:
        logger.error('Failed to parse devices', {'devices_data': devices, 'error': str(e)})
        raise e

File: golang-server/db/devices.go
// db/devices.go
package db

import (
	"database/sql"
	"log"
	"strings"

	"github.com/topper3418/led-hub/golang-server/models"
)

func InitDevices(db *sql.DB) error {
	log.Println("Initializing devices table")
	query, err := ReadSQLInitFile("devices")
	if err != nil {
		return err
	}
	_, err = db.Exec(query)
	if err != nil {
		log.Printf("Failed to initialize devices table: %v", err)
		return err
	}
	return nil
}

func CreateDevice(db *sql.DB, device *models.Device) error {
	log.Printf("Creating device: %+v", device)
	res, err := db.Exec("INSERT INTO devices (mac, ip, name) VALUES (?, ?, ?)", device.Mac, device.IP, device.Name)
	if err != nil {
		log.Printf("Failed to create device: %v", err)
		return err
	}
	id, err := res.LastInsertId()
	if err != nil {
		return err
	}
	device.ID = int(id)
	return nil
}

func UpdateDevice(db *sql.DB, device *models.Device) error {
	log.Printf("Updating device: %+v", device)
	_, err := db.Exec("UPDATE devices SET mac = ?, ip = ?, name = ?, room_id = ? WHERE id = ?", device.Mac, device.IP, device.Name, device.RoomID, device.ID)
	if err != nil {
		log.Printf("Failed to update device: %v", err)
		return err
	}
	return nil
}

func RecordPing(db *sql.DB, deviceID int) error {
	log.Printf("Recording ping for device: %d", deviceID)
	_, err := db.Exec("UPDATE devices SET last_ping = CURRENT_TIMESTAMP WHERE id = ?", deviceID)
	if err != nil {
		log.Printf("Failed to record ping: %v", err)
		return err
	}
	return nil
}

func DeleteDevice(db *sql.DB, deviceID int) error {
	log.Printf("Deleting device: %d", deviceID)
	_, err := db.Exec("DELETE FROM devices WHERE id = ?", deviceID)
	if err != nil {
		log.Printf("Failed to delete device: %v", err)
		return err
	}
	return nil
}

func FindDeviceByID(db *sql.DB, deviceID int) (*models.Device, error) {
	log.Printf("Finding device by id: %d", deviceID)
	device := &models.Device{}
	err := db.QueryRow("SELECT id, mac, name, ip, last_ping, room_id FROM devices WHERE id = ?", deviceID).
		Scan(&device.ID, &device.Mac, &device.Name, &device.IP, &device.LastPing, &device.RoomID)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		log.Printf("Failed to find device by id: %v", err)
		return nil, err
	}
	return device, nil
}

func FindByMac(db *sql.DB, mac string) (*models.Device, error) {
	log.Printf("Finding device by mac: %s", mac)
	device := &models.Device{}
	err := db.QueryRow("SELECT id, mac, name, ip, last_ping, room_id FROM devices WHERE mac = ?", mac).
		Scan(&device.ID, &device.Mac, &device.Name, &device.IP, &device.LastPing, &device.RoomID)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		log.Printf("Failed to find device by mac: %v", err)
		return nil, err
	}
	return device, nil
}

func FindByName(db *sql.DB, name string) (*models.Device, error) {
	log.Printf("Finding device by name: %s", name)
	device := &models.Device{}
	err := db.QueryRow("SELECT id, mac, name, ip, last_ping, room_id FROM devices WHERE name = ?", name).
		Scan(&device.ID, &device.Mac, &device.Name, &device.IP, &device.LastPing, &device.RoomID)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		log.Printf("Failed to find device by name: %v", err)
		return nil, err
	}
	return device, nil
}

func ListDevices(db *sql.DB, room *int, connected *bool) ([]*models.Device, error) {
	log.Printf("Listing devices: room=%v, connected=%v", room, connected)
	query := "SELECT id, mac, name, ip, last_ping, room_id FROM devices"
	var args []interface{}
	var where []string
	if room != nil {
		if *room == 0 {
			where = append(where, "room_id IS NULL")
		} else {
			where = append(where, "room_id = ?")
			args = append(args, *room)
		}
	}
	if connected != nil {
		where = append(where, "connected = ?") // Note: Schema lacks 'connected'; this may error like Python. Derive from last_ping if needed later.
		args = append(args, *connected)
	}
	if len(where) > 0 {
		query += " WHERE " + strings.Join(where, " AND ")
	}
	rows, err := db.Query(query, args...)
	if err != nil {
		log.Printf("Failed to list devices: %v", err)
		return nil, err
	}
	defer rows.Close()
	var devices []*models.Device
	for rows.Next() {
		device := &models.Device{}
		err := rows.Scan(&device.ID, &device.Mac, &device.Name, &device.IP, &device.LastPing, &device.RoomID)
		if err != nil {
			return nil, err
		}
		devices = append(devices, device)
	}
	return devices, nil
}
